mailbox/plugin/quota-mailing/src/main/java/org/apache/james/mailbox/quota/mailing/QuotaMailingListener.java

```diff
@@ -0,0 +1,153 @@
+        this.usersRepository = usersRepository;
+        this.quotaThresholdChangesStore = quotaThresholdChangesStore;
+        this.configuration = QuotaMailingListenerConfiguration.DEFAULT;
+        this.instantSupplier = instantSupplier;
+    }
+
+    public void configure(QuotaMailingListenerConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    public ListenerType getType() {
+        return ListenerType.ONCE;
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return ExecutionMode.SYNCHRONOUS;
+    }
+
+    @Override
+    public void event(Event event) {
+        try {
+            if (event instanceof QuotaUsageUpdatedEvent) {
+                QuotaUsageUpdatedEvent updatedEvent = (QuotaUsageUpdatedEvent) event;
+                User user = getUser(event);
+                handleEvent(updatedEvent, user);
+            }
+        } catch (Exception e) {
+            LOGGER.error("Can not send a quota mail alert", e);
+        }
+    }
+
+    public void handleEvent(QuotaUsageUpdatedEvent updatedEvent, User user) throws UsersRepositoryException, MessagingException {
+        Optional<InformationToEmail> notice = computeInformationToEmail(user, updatedEvent);
+
+        if (notice.isPresent()) {
+            MailAddress sender = mailetContext.getPostmaster();
+            MailAddress recipient = usersRepository.getMailAddressFor(user);
+
+            mailetContext.sendMail(sender, ImmutableList.of(recipient),
+                MimeMessageBuilder.mimeMessageBuilder()
+                    .addFrom(sender.asString())
+                    .addToRecipient(recipient.asString())
+                    .setSubject("Warning: Your email usage just exceeded a configured threshold")
+                    .setText(notice.get().generateReport())
+                    .build());
+        }
+    }
+
+    public Optional<InformationToEmail> computeInformationToEmail(User user, QuotaUsageUpdatedEvent event) {
+        QuotaThreshold countThreshold = configuration.getThresholds().firstExceededThreshold(event.getCountQuota());
+        CompareWithCurrentThreshold compareWithCountThreshold = quotaThresholdChangesStore
+            .retrieveQuotaCountThresholdChanges(user)
+            .compareWithCurrentThreshold(countThreshold, configuration.getGracePeriod());
+
+        QuotaThreshold sizeThreshold = configuration.getThresholds().firstExceededThreshold(event.getSizeQuota());
+        CompareWithCurrentThreshold compareWithCurrentSizeThreshold = quotaThresholdChangesStore
+            .retrieveQuotaSizeThresholdChanges(user)
+            .compareWithCurrentThreshold(sizeThreshold, configuration.getGracePeriod());
+        if (compareWithCountThreshold != CompareWithCurrentThreshold.NO_CHANGES) {
+            Action updateCountThreshold = () -> quotaThresholdChangesStore.persistQuotaCountThresholdChange(user, new QuotaThresholdChange(countThreshold, instantSupplier.now()));
```

Comment:
the method does much more than we its name says.
IMO you do too much in this Listener, you should put the heavy work into a service
BTW, you also mix the feature (mail sending) with threshold passing detection.
We should have a listener that detects thresholds changes to cast an event and another that sends a mail based on this event